helix-core/src/main/java/org/apache/helix/controller/stages/ManagementMessageGenerationPhase.java

```diff
@@ -0,0 +1,129 @@
+import org.apache.helix.model.LiveInstance;
+import org.apache.helix.model.LiveInstance.LiveInstanceStatus;
+import org.apache.helix.model.Message;
+import org.apache.helix.model.PauseSignal;
+import org.apache.helix.util.MessageUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Generates participant status change (freeze/unfreeze) and pending state transition cancellation
+ * messages for management mode pipeline.
+ */
+public class ManagementMessageGenerationPhase extends MessageGenerationPhase {
+  private static final Logger LOG = LoggerFactory.getLogger(ManagementMessageGenerationPhase.class);
+
+  @Override
+  public void process(ClusterEvent event) throws Exception {
+    _eventId = event.getEventId();
+    String clusterName = event.getClusterName();
+    HelixManager manager = event.getAttribute(AttributeName.helixmanager.name());
+    ClusterManagementMode managementMode = event.getAttribute(AttributeName.CLUSTER_STATUS.name());
+    ManagementControllerDataProvider cache =
+        event.getAttribute(AttributeName.ControllerDataProvider.name());
+    if (manager == null || managementMode == null || cache == null) {
+      throw new StageException("Missing attributes in event: " + event
+          + ". Requires HelixManager|ClusterStatus|DataCache");
+    }
+
+    PauseSignal pauseSignal = cache.getPauseSignal();
+    if (cache.getClusterConfig().isStateTransitionCancelEnabled()
+        && pauseSignal != null && pauseSignal.getCancelPendingST()) {
+      // Generate ST cancellation messages.
+      LogUtil.logInfo(LOG, _eventId,
+          "Generating ST cancellation messages for cluster " + clusterName);
+      super.process(event);
+    }
+
+    // Is participant status change still in progress? Create messages
+    if (!ClusterManagementMode.Status.COMPLETED.equals(managementMode.getStatus())) {
+      LogUtil.logInfo(LOG, _eventId, "Generating messages as cluster " + clusterName
+          + " is still in progress to change participant status");
+      List<Message> messages =
+          generateStatusChangeMessages(managementMode.getMode(), cache.getEnabledLiveInstances(),
+              cache.getLiveInstances(), cache.getAllInstancesMessages(), manager.getInstanceName(),
+              manager.getSessionId());
+      MessageOutput messageOutput =
+          event.getAttributeWithDefault(AttributeName.MESSAGES_ALL.name(), new MessageOutput());
+      messageOutput.addStatusChangeMessages(messages);
+      event.addAttribute(AttributeName.MESSAGES_ALL.name(), messageOutput);
+    }
+  }
+
+  private List<Message> generateStatusChangeMessages(ClusterManagementMode.Type managementMode,
+      Set<String> enabledLiveInstances, Map<String, LiveInstance> liveInstanceMap,
+      Map<String, Collection<Message>> allInstanceMessages, String managerInstance,
+      String managerSessionId) {
+    List<Message> messagesToSend = new ArrayList<>();
+
+    LiveInstanceStatus fromStatus = null;
```

Comment:
I would suggest you do the similar logic as message generation.  Use the current status of instance as "FROMSTATE" and expected state as "TOSTATE". Then check whether instance already in that state.

This logic could be more generic.